### DISCUSSION TITLE
Debugging OAuth2 tokens?

### DISCUSSION THREAD

**Discussion by lf-** [kanidm/kanidm#2783] [Q&A]
Hi! I'm integrating a forgejo instance with kanidm and I'm struggling to get the username mapping to work properly (namely it should be just the local part of the username, instead of something at idm.example.com). I'm not sure if it's broken because of my forgejo config or because the info is literally not in the token, and I have no idea how to actually *obtain* "a token" as a user to find that out. 

In keycloak I just bap some buttons in the evaluate tab of the client and immediately get any of the token JSON I need. As far as I can tell there's no way to get access tokens out of kanidm for debugging to debug why apps are behaving weirdly, short of building a fake OAuth2 app. 

Is there a fake OAuth2 app that works (I tried https://oauthdebugger.com/ and its PKCE support or UI seems to sufficiently not work that I don't get a token out of it), or some kanidm command I've not found?

**Comment by yaleman** [comment]
There's the `prefer-short-username` option for the OAuth2 client which will set that instead of the full SPN in the userinfo, which should sort the issue you're trying to solve 😄 


As far as debugging, from memory oauth2-proxy has some debug options I think.

**Reply by lf-** [reply]
I later also found a Google OAuth2 toolbox which looks plausible.

**Reply by lf-** [reply]
https://developers.google.com/oauthplayground/#step3&scopes=email%2C%20profile%2C%20openid&url=https%3A%2F%2F&content_type=application%2Fjson&http_method=GET&useDefaultOauthCred=checked&oauthEndpointSelect=Custom&oauthAuthEndpointValue=https%3A%2F%2Fidm.example.com%2Fui%2Foauth2%3Fstate%3Dmeowing&oauthTokenEndpointValue=https%3A%2F%2Fidm.example.com%2Foauth2%2Ftoken&includeCredentials=unchecked&accessTokenType=bearer&autoRefreshToken=unchecked&accessType=offline&prompt=consent&response_type=code&wrapLines=on

man this software is awful. but it technically does work if you disable PKCE.

**Reply by lf-** [reply]
I would say this use case is still a nightmare that really needs far better support in kanidm given the state of fake oauth services on the internet, but at least it's solvable now with this site and https://jwt.io.

Here is an id token.

```json
{
  "iss": "https://idm.example.com/oauth2/openid/temp2_oauth",
  "sub": "eeeeeeeee-eeee-eeee-eeee-eeeeeeeeeeee",
  "aud": "temp2_oauth",
  "exp": 1716237671,
  "nbf": 1716236771,
  "iat": 1716236771,
  "azp": "temp2_oauth",
  "name": "jade",
  "preferred_username": "jade@idm.example.com",
  "email": "redacted",
  "email_verified": true,
  "scopes": [
    "email",
    "openid",
    "profile"
  ]
}
```

**Reply by yaleman** [reply]
Is this the bappable keycloak button you're talking about? Seems like a neat feature.

https://www.keycloak.org/docs/latest/server_admin/index.html#_client_scopes_evaluate

<img width="995" alt="Screenshot 2024-05-21 at 08 34 18" src="https://github.com/kanidm/kanidm/assets/168188/47455d38-947d-431d-ae1e-bfc8fed0cf32">

**Reply by lf-** [reply]
yup, this exact one.
